common/lru/blob_lru.go

```diff
@@ -19,41 +19,40 @@ package lru
 import (
 	"math"
 	"sync"
-
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/hashicorp/golang-lru/simplelru"
 )
 
-// SizeConstrainedLRU is a wrapper around simplelru.LRU. The simplelru.LRU is capable
-// of item-count constraints, but is not capable of enforcing a byte-size constraint,
-// hence this wrapper.
+// blobType is the type constraint for values stored in SizeConstrainedCache.
+type blobType interface {
+	~[]byte | ~string
+}
+
+// SizeConstrainedCache is a cache where capacity is in bytes (instead of item count). When the cache
+// is at capacity, and a new item is added, older items are evicted until the size
+// constraint is met.
+//
 // OBS: This cache assumes that items are content-addressed: keys are unique per content.
 // In other words: two Add(..) with the same key K, will always have the same value V.
-type SizeConstrainedLRU struct {
+type SizeConstrainedCache[K comparable, V blobType] struct {
 	size    uint64
 	maxSize uint64
-	lru     *simplelru.LRU
+	lru     BasicLRU[K, V]
 	lock    sync.Mutex
 }
 
-// NewSizeConstrainedLRU creates a new SizeConstrainedLRU.
-func NewSizeConstrainedLRU(max uint64) *SizeConstrainedLRU {
-	lru, err := simplelru.NewLRU(math.MaxInt, nil)
-	if err != nil {
-		panic(err)
-	}
-	return &SizeConstrainedLRU{
+// NewSizeConstrainedCache creates a new size-constrained LRU cache.
+func NewSizeConstrainedCache[K comparable, V blobType](maxSize uint64) *SizeConstrainedCache[K, V] {
+	return &SizeConstrainedCache[K, V]{
 		size:    0,
-		maxSize: max,
-		lru:     lru,
+		maxSize: maxSize,
+		lru:     NewBasicLRU[K, V](math.MaxInt),
 	}
 }
 
 // Add adds a value to the cache.  Returns true if an eviction occurred.
 // OBS: This cache assumes that items are content-addressed: keys are unique per content.
 // In other words: two Add(..) with the same key K, will always have the same value V.
 // OBS: The value is _not_ copied on Add, so the caller must not modify it afterwards.
-func (c *SizeConstrainedLRU) Add(key common.Hash, value []byte) (evicted bool) {
+func (c *SizeConstrainedCache[K, V]) Add(key K, value V) (evicted bool) {
```

Comment:
One thing that I feel should be checked, is the behaviour with `nil` values. Right now, we never store `nil` nor `empty` code in the codecache, but that might change. 

So basically, how does the `SizeConstrainedCache` cache behave with regards to `nil` vs `0-length` byte arrays. (And how do we want it to behave?)

I'm just writing this down here so it isn't forgotten

Reply:
```golang
// TestBlobEmpties tests that empty/nil values are handled correctly.
func TestBlobEmpties(t *testing.T) {
	lru := NewBlobLRU[common.Hash, []byte](100)
	// This test abuses the lru a bit, using different keys for identical value(s).
	for i := 0; i < 10; i++ {
		lru.Add(common.Hash{byte(i)}, []byte{})
		lru.Add(common.Hash{byte(255 - i)}, nil)
	}
	// The size should not count, only the values count. So this could be a DoS
	// since it basically has no cap, and it is intentionally overloaded with
	// different-keyed 0-length values.
	if have, want := lru.size, uint64(0); have != want {
		t.Fatalf("size wrong, have %d want %d", have, want)
	}
	for i := 0; i < 10; i++ {
		if v, ok := lru.Get(common.Hash{byte(i)}); !ok {
			t.Fatalf("test %d: expected presence", i)
		} else if v == nil {
			t.Fatalf("test %d, v is nil", i)
		}
		if v, ok := lru.Get(common.Hash{byte(255 - i)}); !ok {
			t.Fatalf("test %d: expected presence", i)
		} else if v != nil {
			t.Fatalf("test %d, v is not nil", i)
		}
	}
}
```
We could use this test. It even passes :)

Reply:
Added the test.